core/src/main/java/org/elasticsearch/action/admin/cluster/node/tasks/cancel/TransportCancelTasksAction.java

```diff
@@ -118,12 +120,44 @@ protected synchronized void taskOperation(CancelTasksRequest request, Cancellabl
         Set<String> childNodes = taskManager.cancel(cancellableTask, request.getReason(), banLock::onTaskFinished);
         if (childNodes != null) {
             if (childNodes.isEmpty()) {
+                // The task has no child tasks, so we can return immediately
                 logger.trace("cancelling task {} with no children", cancellableTask.getId());
                 listener.onResponse(cancellableTask.taskInfo(clusterService.localNode().getId(), false));
             } else {
+                // The task has some child tasks, we need to wait for until ban is set on all nodes
                 logger.trace("cancelling task {} with children on nodes [{}]", cancellableTask.getId(), childNodes);
-                setBanOnNodes(request.getReason(), cancellableTask, childNodes, banLock);
-                listener.onResponse(cancellableTask.taskInfo(clusterService.localNode().getId(), false));
+                String nodeId = clusterService.localNode().getId();
+                AtomicInteger responses = new AtomicInteger(childNodes.size());
+                List<Exception> failures = new ArrayList<>();
```

Comment:
I feel like we implement this pattern enough times that we should make a helper for it at some point. No need now, but at some point.